reactive/kotlinx-coroutines-rx2/src/RxConvert.kt

```diff
@@ -81,7 +81,12 @@ public fun <T: Any> ObservableSource<T>.asFlow(): Flow<T> = callbackFlow {
     val observer = object : Observer<T> {
         override fun onComplete() { close() }
         override fun onSubscribe(d: Disposable) { if (!disposableRef.compareAndSet(null, d)) d.dispose() }
-        override fun onNext(t: T) { sendBlocking(t) }
+        override fun onNext(t: T) {
+            try {
+                sendBlocking(t)
+            } catch (ignored: Exception) { //TODO: Replace when this issue is fixed: https://github.com/Kotlin/kotlinx.coroutines/issues/974
```

Comment:
Please add a corresponding unit-test (both to rx2 and rx3) that fails if this change is reverted.

Reply:
Adding a test for this kind of race condition is not trivial… I mean, I have no clue of a way to implement it reliably.

If you know a way, please commit it to this branch, the "Allow edit from maintainers" checkbox is checked.

Reply:
For race conditions like this, we are writing stress tests that fail at least sometimes and run them nightly with increased duration. We name them ending with `StressTest`. You can check out examples.

Reply:
Been able to repro the consequence of the throwing `sendBlocking` via this stress test: https://gist.github.com/drinkthestars/4fa82b744a623612b481c05efeacd812

The crux of the test is this:
``` Kotlin
val outerObservable = Observable.interval(1L, TimeUnit.MILLISECONDS)
val innerObservable = Observable.interval(1L, TimeUnit.MILLISECONDS)

val job = launch(newSingleThreadContext("Collector Thread")) {
   outerObservable
      .asFlow()
      .flatMapLatest { // The collect cancellations caused by this trigger the race condition
         innerObservable.asFlow()
      }
      .collect { delay(200) }
}
```

Reply:
@drinkthestars Would you send a PR with that stress test at the right location against the branch of this PR in my fork?

I'll then merge it so it lands in this PR, and keeps the attribution to you.

Reply:
Will do.

Reply:
@LouisCAD Check out https://github.com/LouisCAD/kotlinx.coroutines/pull/1. Let me know if there are any issues, thanks!

Reply:
LGTM but I didn't test locally.

Forwarding that without the fix of this PR, these stress tests I just merged would fail somewhere between the iterations 1 and 7.